File: src/Trie.java
public class Trie {
    Node root;

    /**
     * Initialize your data structure here.
     */
    public Trie() {
        root = new Node();
    }

    /**
     * Inserts a word into the trie.
     */
    public void insert(String word) {
        Node node = root;
        for (int i = 0; i < word.length(); i++) {
            int currentChar = word.charAt(i) - 'a';
            if (node.children[currentChar] == null) {
                node.children[currentChar] = new Node();
            }
            node = node.children[currentChar];
        }
        node.end = true;
    }

    /**
     * Returns if the word is in the trie.
     */
    public boolean search(String word) {
        Node node = root;
        for (int i = 0; i < word.length(); i++) {
            if (node == null)
                return false;
            node = node.children[word.charAt(i) - 'a'];
        }
        return node != null && node.end;
    }

    /**
     * Returns if there is any word in the trie that starts with the given prefix.
     */
    public boolean startsWith(String prefix) {
        Node node = root;
        for (int i = 0; i < prefix.length(); i++) {
            if (node == null)
                return false;
            node = node.children[prefix.charAt(i) - 'a'];
        }
        return node != null;
    }

    class Node {
        Node[] children;
        boolean end;

        public Node() {
            children = new Node[26];
            end = false;
        }
    }
}

File: src/utils/Trie.py
from typing import Set

class TrieNode:
    def __init__(self):
        self.next = {}
        self.words: Set[str] = set()

class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def add(self, word: str):
        self.root.words.add(word)
        cur = self.root
        for c in word:
            if c not in cur.next:
                cur.next[c] = TrieNode()
            cur = cur.next[c]
            cur.words.add(word)
    
    def search(self, word: str) -> Set[str]:
        cur = self.root
        for c in word:
            if c not in cur.next:
                return set()
            cur = cur.next[c]
        return cur.words